sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/v2/ResolveMultipartRelationSuite.scala

```diff
@@ -0,0 +1,98 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.catalog.v2
+
+import org.apache.spark.sql.catalog.v2.{CatalogNotFoundException, CatalogPlugin}
+import org.apache.spark.sql.catalyst.CatalogIdentifier
+import org.apache.spark.sql.catalyst.analysis.{AnalysisTest, Analyzer, EliminateSubqueryAliases, NamedRelation}
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.parser.CatalystSqlParser
+import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+private case class TestUnresolvedMultipartRelation(parts: Seq[String])
+    extends LeafNode with NamedRelation {
+
+  override def name: String = "TestUnresolvedMultipartRelation"
+
+  override def output: Seq[Attribute] = Nil
+
+  override lazy val resolved = false
+}
+
+private case class TestMultipartRelation(catalog: Option[CatalogPlugin], ident: CatalogIdentifier)
+    extends LeafNode with NamedRelation {
+
+  override def name: String = "TestMultipartRelation"
+
+  override def output: Seq[Attribute] = Nil
+
+  override lazy val resolved = true
+}
+
+private case class TestMultipartAnalysis(analyzer: Analyzer) extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators  {
+
+    case TestUnresolvedMultipartRelation(analyzer.CatalogRef(catalog, ident)) =>
+      TestMultipartRelation(catalog, ident)
+  }
+}
+
+private class TestCatalogPlugin(override val name: String) extends CatalogPlugin {
+
+  override def initialize(name: String, options: CaseInsensitiveStringMap): Unit = Unit
+}
+
+class ResolveMultipartRelationSuite extends AnalysisTest {
+  import CatalystSqlParser._
+
+  private val analyzer = makeAnalyzer(caseSensitive = false)
+
+  private val catalogs = Seq("prod", "test").map(name => name -> new TestCatalogPlugin(name)).toMap
+
+  private def lookupCatalog(catalog: String): CatalogPlugin =
+    catalogs.getOrElse(catalog, throw new CatalogNotFoundException("Not found"))
+
+  private def makeAnalyzer(caseSensitive: Boolean) = {
+    val conf = new SQLConf().copy(SQLConf.CASE_SENSITIVE -> caseSensitive)
+    new Analyzer(Some(lookupCatalog _), null, conf) {
+      override val extendedResolutionRules =
+        EliminateSubqueryAliases :: TestMultipartAnalysis(this) :: Nil
```

Comment:
This uses a lot of temporary classes to simulate future rules that match multi-part identifiers. I think I would rather include an update that adds new `UnresolvedRelation` nodes and uses them instead of test plan nodes, but I'd be interested to hear whether @cloud-fan agrees.

Reply:
OK either way. I have already converted SELECT/INSERT/DROP code paths to support multi-catalog in my private 2.3 branch. Pretty straightforward. Converting CREATE would be a lot easier with Ryan's PR 24029.